File: packages/simulation/src/engine/indicators.ts
/**
 * Technical Indicators
 *
 * Pure implementations matching Python indicators.py
 * No I/O, no dependencies on external packages.
 */

/**
 * Calculate EMA (Exponential Moving Average)
 *
 * @param values - Array of close prices
 * @param period - EMA period (must be > 0)
 * @returns Array of EMA values (null for indices < period - 1)
 */
export function ema(values: number[], period: number): (number | null)[] {
  if (period <= 0) {
    throw new Error('EMA period must be > 0');
  }

  const out: (number | null)[] = new Array(values.length).fill(null);

  if (values.length === 0) {
    return out;
  }

  const k = 2 / (period + 1);
  let emaPrev: number | null = null;

  for (let i = 0; i < values.length; i++) {
    if (i < period - 1) {
      continue; // Already null
    }

    if (i === period - 1) {
      // Seed with SMA
      const sma = values.slice(0, period).reduce((sum, v) => sum + v, 0) / period;
      emaPrev = sma;
      out[i] = emaPrev;
      continue;
    }

    if (emaPrev === null) {
      throw new Error('EMA calculation error: emaPrev is null');
    }

    emaPrev = (values[i] - emaPrev) * k + emaPrev;
    out[i] = emaPrev;
  }

  return out;
}

/**
 * Calculate RSI (Relative Strength Index)
 *
 * Uses Wilder smoothing method.
 *
 * @param close - Array of close prices
 * @param period - RSI period (must be > 0)
 * @returns Array of RSI values (null for indices < period)
 */
export function rsi(close: number[], period: number): (number | null)[] {
  if (period <= 0) {
    throw new Error('RSI period must be > 0');
  }

  const n = close.length;
  const out: (number | null)[] = new Array(n).fill(null);

  if (n === 0) {
    return out;
  }

  const gains: number[] = new Array(n).fill(0);
  const losses: number[] = new Array(n).fill(0);

  for (let i = 1; i < n; i++) {
    const ch = close[i] - close[i - 1];
    gains[i] = Math.max(ch, 0.0);
    losses[i] = Math.max(-ch, 0.0);
  }

  if (n <= period) {
    return out;
  }

  // Wilder smoothing - initial average
  let avgGain = gains.slice(1, period + 1).reduce((sum, v) => sum + v, 0) / period;
  let avgLoss = losses.slice(1, period + 1).reduce((sum, v) => sum + v, 0) / period;

  const calcRsi = (ag: number, al: number): number => {
    if (al === 0) {
      return 100.0;
    }
    const rs = ag / al;
    return 100.0 - 100.0 / (1.0 + rs);
  };

  out[period] = calcRsi(avgGain, avgLoss);

  // Continue with Wilder smoothing
  for (let i = period + 1; i < n; i++) {
    avgGain = (avgGain * (period - 1) + gains[i]) / period;
    avgLoss = (avgLoss * (period - 1) + losses[i]) / period;
    out[i] = calcRsi(avgGain, avgLoss);
  }

  return out;
}

File: strategy-ui/app/services/indicators.py
from __future__ import annotations

from typing import List, Optional

def ema(values: List[float], period: int) -> List[Optional[float]]:
    if period <= 0:
        raise ValueError("EMA period must be > 0")
    out: List[Optional[float]] = [None] * len(values)
    if len(values) == 0:
        return out
    k = 2 / (period + 1)
    ema_prev: Optional[float] = None
    for i, v in enumerate(values):
        if i < period - 1:
            out[i] = None
            continue
        if i == period - 1:
            # seed with SMA
            sma = sum(values[:period]) / period
            ema_prev = sma
            out[i] = ema_prev
            continue
        assert ema_prev is not None
        ema_prev = (v - ema_prev) * k + ema_prev
        out[i] = ema_prev
    return out

def rsi(close: List[float], period: int) -> List[Optional[float]]:
    if period <= 0:
        raise ValueError("RSI period must be > 0")
    n = len(close)
    out: List[Optional[float]] = [None] * n
    if n == 0:
        return out

    gains: List[float] = [0.0] * n
    losses: List[float] = [0.0] * n
    for i in range(1, n):
        ch = close[i] - close[i-1]
        gains[i] = max(ch, 0.0)
        losses[i] = max(-ch, 0.0)

    # Wilder smoothing
    avg_gain = sum(gains[1:period+1]) / period if n > period else 0.0
    avg_loss = sum(losses[1:period+1]) / period if n > period else 0.0

    if n <= period:
        return out

    def calc_rsi(ag: float, al: float) -> float:
        if al == 0:
            return 100.0
        rs = ag / al
        return 100.0 - (100.0 / (1.0 + rs))

    out[period] = calc_rsi(avg_gain, avg_loss)

    for i in range(period + 1, n):
        avg_gain = (avg_gain * (period - 1) + gains[i]) / period
        avg_loss = (avg_loss * (period - 1) + losses[i]) / period
        out[i] = calc_rsi(avg_gain, avg_loss)

    return out

